File: psc/2014/gnyrc 2014/h.java
import java.math.BigInteger;
import java.util.Scanner;

class h {
	static void fuck(int t, BigInteger n) {
		BigInteger TWO = new BigInteger("2");
		BigInteger a = BigInteger.ZERO, b = BigInteger.ONE, c = BigInteger.ONE, d = new BigInteger(
				n.toString());
		BigInteger num = BigInteger.ZERO, de = BigInteger.ONE;
		BigInteger lastb = BigInteger.ONE, currentb;
		while (c.compareTo(n) <= 0) {
			BigInteger k = n.add(b).divide(d);
			BigInteger ta = a, tb = b;
			a = c;
			b = d;
			c = k.multiply(c).subtract(ta);
			d = k.multiply(d).subtract(tb);
			currentb = b;
			BigInteger tn = lastb.multiply(lastb).add(
					currentb.multiply(currentb));
			BigInteger td = lastb.multiply(currentb);
			num = num.multiply(td).add(de.multiply(tn));
			de = de.multiply(td);
			lastb = currentb;
			if (b.equals(a.multiply(TWO))) {
				break;
			}
		}
		num=num.divide(num.gcd(de));
		System.out.println(t+" "+num+"/2");
	}

	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int tc = sc.nextInt();
		for (int t = 1; t <= tc; t++) {
			BigInteger x;
			x = sc.nextBigInteger();
			x = sc.nextBigInteger();
			fuck(t, x);
		}
	}
}

File: psc/2014/gnyrc 2014/h.c

__uint128_t gcd(__uint128_t a, __uint128_t b) {
	if (b == 0) {
		return a;
	} else {
		return gcd(b, a % b);
	}
}

void ughhhhh(int t,__uint128_t n) {
	__uint128_t a = 0, b = 1, c = 1, d = n;
	__uint128_t num = 0, de = 1;
	__uint128_t lastb = 1, currentb;
	while (c <= n) {
		__uint128_t k = ((n + b) / d);
		__uint128_t ta = a, tb = b;
		a = c;
		b = d;
		c = k * c - ta;
		d = k * d - tb;
		currentb = b;
		
		__uint128_t tn = lastb * lastb + currentb * currentb;
		__uint128_t td = lastb * currentb;
		
		num = num * td + de * tn;
		de *= td;

		__uint128_t g=gcd(num,de);
		num/=g;
		de/g

		lastb = currentb;
		if (b == a * 2) {
			break;
		}
	}
	printf("%d %d/2\n",t,num);
}

int main() {
	int tc;
	scanf("%d",&tc);
	int t=1;
	while(t <= tc) {
		__uint128_t x;
		scanf("%d",&x);
		scanf("%d",&x);
		ughhhhh(t,x);
		t++;
	}
}
